cmd/clusterctl/client/cluster/cert_manager.go

```diff
@@ -327,7 +328,7 @@ func (cm *certManagerClient) shouldUpgrade(objs []unstructured.Unstructured) (st
 			return "", "", false, errors.Wrapf(err, "failed to parse version for cert-manager component %s/%s", obj.GetKind(), obj.GetName())
 		}
 
-		c, err := objSemVersion.Compare(config.Version())
+		c, err := objSemVersion.Compare(configVersion)
```

Comment:
Have you considered to use `util.Compare(a, b, WithBuildTags())` instead?
@ykakarap opinions ^^

Reply:
I just tried that. The problem is that method uses string comparison to determine precedence. That works if you use more normalized build metadata schemes like `v0.0.1+build.1` -> `v0.0.1+build.2`. But it won't work with others, like using a commit hash `v0.0.1+hf87h89` -> `v0.0.1+hf87h88`

I was aiming for something more simple that supports the latter

Reply:
@g-gaston The only case where this PR and using `util.Compare(a, b, WithBuildTags())` will defer will be if the new build tag is ordered higher in string compare but you still want `needUpgrade` to be set to `true` (because it changed).

If you want to achieve the above case then current approach is the way to go 👍🏼

Reply:
@ykakarap yeah, I couldn't have said better. Thanks!

Reply:
What about adding an WithUnsortedBuildTags option?
I really would like to keep version compare logic in a single place given that we just completed a cleanup effort for making this consistent across the codebase

Reply:
Should build metadata even be considered sortable? By default build information isn't considered sortable, what would be the downside to only catch difference in build metadata rather than having an ordered comparison?

Reply:
@fabriziopandini 
I can add that option. What would be the behavior: if normal `Compare` returns 0 and build tags differ, always return -1?

However, I think I agree with @vincepri and I find a bit confusing taking build metadata into account for an ordered comparison.

If the goal is to keep all the semver logic in the `version` package, maybe the solution comes by adding a new method like `version.IsNewerOrSameAndDifferentMetadata`? (pls advice for a better name if you an idea). At the end, we don't need an ordered comparison in this scenario, we just need a true/false for this specific condition.

Reply:
Sorted metadata is a requirement from folks using versioning schema like v0.0.1+build.1 -> v0.0.1+build.2, which I kind of agree it is a stretch, but we can't drop this without breaking users expectations.

I prefer a single compare method with many explicit options to having different methods; however, I'm happy to consider better options names if this helps to make things more clear

Reply:
Makes sense, thanks!

Reply:
One note though is that now we're behaving in different ways in different places, is there a way that we could do a smarter detection if strings shouldn't be compared with `>`, but if should just be different?

Reply:
not sure.
TBH now that I think about this different use cases, there is a possible conflict between this changeset and requirement from folks using versioning schema like v0.0.1+build.1 -> v0.0.1+build.2.

If they apply the same versioning schema to cert manager (and I'm pretty sure there are users doing this), they will expect also cert manager upgrade to consider sorted metadata to decide when to upgrade - a thing that is not implemented now, but it seems a sort of bug now that I think about it -

This PR instead proposes to just check for metadata equality, without sorting, but this will give unexpected results for people using v0.0.1+build.1 naming with cert manager.

My first reaction to this realisation is that we should use sorted metadata also here, so at least we are consistent, but as downside, we should somehow document that tagging custom-built components with hashes is not supported...

@vincepri @ykakarap opinions?

Reply:
I might be misinterpreting this but here's a quote from sem ver 2.0.0:
> Build metadata MUST be ignored when determining version precedence. Thus two versions that differ only in the build metadata, have the same precedence. Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85, 1.0.0+21AF26D3—-117B344092BD.
> [...]
> Precedence MUST be calculated by separating the version into major, minor, patch and pre-release identifiers in that order (Build metadata does not figure into precedence).
> https://semver.org/

Reply:
> Sounds like we violate the semver spec when we support ordering by build metadata.

Kind of, but this was accepted given the requirements from folks using versioning schema like v0.0.1+build.1 -> v0.0.1+build.2 and we are already supporting this all across the codebase.
Given that IMO it is a bug we are not supporting the extended version precedence rule also here.

Reply:
Agree. We shouldn't be inconsistent here when we intentionally did it in another parts of the code base. I wasn't aware of that.

Reply:
Ok, got it. With all this new info/requirements, I have a different proposal:
* Use `util.Compare(a, b, WithBuildTags())` for cert-manager `shouldUpgrade`
* When comparing `buildIdentifier`s in the `compare` method, if both are strings:
  * If equal, return 0
  * If not equal, return -1

With the proposed solution:
* We still support `v0.0.1+build.1` -> `v0.0.1+build.2` usecase (it will behave the same way it does now)
* We support hashes in the build metadata
* We keep all the semver compare logic in one place (the `version` package)
* The only thing that changes is if folks try to do something like `v0.0.1+build2` -> `v0.0.1+build1`. That would return -1 with the proposed solution, when right now returns 1. But that doesn't seem like a usecase that needs to be supported

If it seems to risky to change the behavior of `WithBuildTags()`, we can create a different option

Reply:
How about supporting comparison strategies in `WithBuildTags()`? So the option `WithBuildTags()` becomes `WithBuildTags(strategy string)` the list of supported strategies are `AlphaNumeric` and `Random`. 
- `AlphaNumeric` should be the default strategy (the strategy we currently use). 
- `Random` will return -1 as long as the build tags are different.

Reply:
That sounds reasonable. However, I were to use the `Random` strategy, cert-manager wouldn't support the `v0.0.1+build.1` -> `v0.0.1+build.2` usecase, and it seemed like @fabriziopandini was very inclined to support it here as well. In fact, it would support it, but it would also support `v0.0.1+build.2` -> `v0.0.1+build.1`, which I believe is unwanted.

Reply:
You are right, I don't believe we should support different strategies. From @fabriziopandini comments it looks like we want to be consistent across our code base in how we compare versions. Within our codebase we currently use  build tags as well to compare versions (`util.Compare(a, b, WithBuildTags())`) and we should continue doing it.

I am +1 with @fabriziopandini's comment that we should be consistent in the version compare method we use everywhere and document that custom-built components with hashes is not supported.
>My first reaction to this realisation is that we should use sorted metadata also here, so at least we are consistent, but as downside, we should somehow document that tagging custom-built components with hashes is not supported...


This would mean cert-manger will also use the existing `WithBuildTags()` way of doing things.

Reply:
Just for my own understanding, what's the reason to perform string ordered comparison right now?
I totally understand the `v0.0.1+build.1` -> `v0.0.1+build.2` usecase, but that doesn't require it, it only requires ordered comparison for the numbers
I'm wondering if there is an actual usecase to support that implementation or if it is something that can potentially be changed elsewhere, achieving both consistency across the codebase and hash support

Reply:
@ykakarap yeah, exactly!
I'll go ahead and implement the changes I'm proposing and we can do another round of discussions over them
It'll probably be easier that way :)

Reply:
@ykakarap done!